tfhe/src/integer/server_key/radix_parallel/shift.rs

```diff
@@ -268,10 +367,103 @@ impl ServerKey {
+
+        if shift > total_num_bits {
+            for block in ct.blocks.iter_mut() {
+                self.key.create_trivial_assign(block, 0)
+            }
+            return;
+        }
+
+        // rotate right as the blocks are from LSB to MSB
+        ct.blocks.rotate_right(rotations);
+        // Every block below 'rotations' should be discarded
+        for block in &mut ct.blocks[..rotations] {
+            self.key.create_trivial_assign(block, 0)
+        }
+
+        if shift_within_block == 0 || rotations == ct.blocks.len() {
+            return;
+        }
+
+        // Since we require that carries are empty,
+        // we can use the bivariate bps to shift and propagate in parallel at the same time
+        // instead of first shifting then propagating
+        //
+        // The first block is done separately as it does not
+        // need to recuperate the shifted bits from its previous block,
+        // and also that way is does not need a special case for when rotations == 0
+        let create_blocks_using_bivariate_pbs = || {
+            let lut = self
+                .key
+                .generate_accumulator_bivariate(|previous_block, current_block| {
+                    let current_block = current_block << shift_within_block;
+                    let previous_block = previous_block << shift_within_block;
+
+                    let message_of_current_block =
+                        current_block % self.key.message_modulus.0 as u64;
+                    let carry_of_previous_block =
+                        previous_block / self.key.message_modulus.0 as u64;
+                    message_of_current_block + carry_of_previous_block
+                });
+            let partial_blocks = ct.blocks[rotations..]
+                .par_windows(2)
+                .map(|blocks| {
+                    let (previous_block, current_block) = (&blocks[0], &blocks[1]);
+                    self.key.unchecked_apply_lookup_table_bivariate(
+                        previous_block,
+                        current_block,
+                        &lut,
+                    )
+                })
+                .collect::<Vec<_>>();
+            partial_blocks
+        };
+
+        let shift_last_block = || {
+            let mut block = ct.blocks[rotations].clone();
+            self.key
+                .scalar_left_shift_assign(&mut block, shift_within_block as u8);
+            block
+        };
+
+        let (partial_blocks, block) =
+            rayon::join(create_blocks_using_bivariate_pbs, shift_last_block);
+
+        // We started with num_blocks, discarded 'rotations' blocks
+        // and did the last one separately
+        let expected_num_partial_block = ct.blocks.len() - rotations - 1;
+        assert_eq!(partial_blocks.len(), expected_num_partial_block);
+        ct.blocks[rotations] = block;
+        for (block, shifted_block) in ct.blocks[rotations + 1..]
```

Comment:
I still believe something like

let dest_blocks = &mut ct.blocks[rotations + 1..];
assert_eq!(partial_blocks.ln(), dest_blocks.len());
for ... in dest_blocks.iter_mut()...

or using izip!/zip checked would be more readable 🤔